### DISCUSSION TITLE
Pybricks 3.0 Release Candidate out now. Please help us test!

### DISCUSSION THREAD

**Discussion by laurensvalk** [pybricks/support#334] [Announcements]
We are super excited that we are finally getting ready for the official Pybricks 3.0 release!

That's right, no more *beta*, but a solid and stable release.

This release supports MicroPython coding for:
- The LEGO Technic Hub.
- The LEGO City Hub.
- The LEGO BOOST Move Hub.
- All motors and sensors, including SPIKE and MINDSTORMS motors and sensors.

## We need your help!

How, you ask? Well:

# :arrow_right: :arrow_right:  **[Start coding, find bugs :bug: ! ](https://www.code.pybricks.com/)** :arrow_left: :arrow_left: 

If you want to help, here's a few things to look out for especially:

- **Firmware installation issues:**
   - Try it on Windows, Mac, or Linux
   - Try it on your latest Laptop and on your grandma's PC.
   - Try it with Google Chrome and Microsoft Edge.
   - Try including a program in the firmware via the settings menu.
- **Running program issues:**
   - Try super short programs or really large ones.
   - Try programs that print a lot, or nothing at all.
   - Try programs that use a lot of memory.
   - Try programs that run for a very long time.
   - Try to start and stop programs often.
   - Try to disconnect from one computer, then continue coding on another.
- **Hub, motor, and sensor issues:**
   - Try as many sensors, motors, and hubs as you have.
   - Try using lots of motors and sensors at the same time.
   - Try doing a lot with motors and sensors in fast loops.
   - Try low batteries and full batteries.
- **Documentation issues:**
   - Try as many of the included [example snippets](https://www.code.pybricks.com/) as you can.
   - Please stick to the documentation inside the coding app for now. While there's more on our [docs site](https://docs.pybricks.com/), that's reserved for a future release. Thanks!
   - Please share (or fix!) any typo or broken link that you can find.
- **Issues with the Pybricks Code editor:**
   - Try all the buttons, bells, and whistles.
   - Try the REPL for interactive coding.
   - Try saving and loading code files from your computer.
   - Try it in a website tab, and try installing it as an app via the settings menu.

Be sure to get creative and try to break as much as you can. In fact, most issues are probably not even on this list!

And remember, the sooner we complete this 3.0 release, the sooner we can work on the 3.1 release which supports the SPIKE Prime Hub and MINDSTORMS Robot inventor Hub.

## Please tell us about your findings!

Please let us know about any errors or unexpected results. That way we can fix them before the final release.

Feel free to just comment here, even if everything is just working as expected. We want to hear that too!

You may also open a new [issue](https://github.com/pybricks/support/issues) directly.

## A big thanks to our users, contributors and sponsors!

We want to say *THANK YOU* to all our users, contributors, and [sponsors](https://github.com/sponsors/pybricks). You've been a great help!

As a sign of appreciation, we'd like to invite you to a video call to discuss your findings and future ideas for Pybricks. We'd love to see your creations with Pybricks too.

We don't have all of your email addresses, so if you'd like to participate, please drop us an email as follows:
- To: team@pybricks.com
- Title: Pybricks call
- Body: Your GitHub user name

Thanks especially to our contributors and public sponsors, listed in alphabetical order (apologies if I've missed your name!):
@AntoniLuongPham, @ArianaLuongPham, @BertLindeman, @codeadamca, @DimitriDekvere, @fkleon, @JanderII, @Jason733i, @JorgePe, @littldr, @martynboogaarts, @mwinkler, @Novakasa, @NStrijbosch, @PhoBricks, @thesynman, @TheVinhLuong102, @VegaMansk, @Vinz1911, @yipjustin, @ZPhilo

**Comment by scotus-49** [comment]
### It Works on Raspberry Pi 4 - a quick test

Firstly, I must offer my personal congratulations to you both for **a truly outstanding achievement**. Well done to you both!! 

I hope that you can inspire new generations of **_Desktop Robotics Designers_** to experiment, explore and learn. The hardware and software basics are now in place.

I read your call for feedback, and fired up my Raspberry Pi 4 to check how the new software was working on this popular maker platform. Here is my (quick) mini-report:

System: Raspbian GNU/Linux | Version 10  (Buster) |  Linux version 5.10.17-v7l+ 
Browser:  Chromium Version 90.0.4430.93 (Developer Build - raspi OS standard distro browser) 
Pybricks Code Version: Pybricks Code v1.0.0-rc.2

**Hubs**: MoveHub |  InventorHub
**Motors**: OK - work as expected
**Sensors**: not yet checked
**Bluetooth**:  No connection problems experienced. 

**MoveHub Firmware update**: the update lasted about 7 min, and the first two upgrade attempts did not work. I kept getting a red warning box coming up - _GATT Error Unknown_ - at the end of the update. However,  after the third try, despite not getting an **OK success** notification (only the familiar red warning window), running code to move the inbuilt motors worked as expected.
With my Inventor hub, the motors reacted to the downloaded code OK. 
I will check sensors etc when I have a bit more time.

Regards from Berlin,
Francis Sim

**Reply by dlech** [reply]
This is expected on Debian Buster since it has BlueZ 5.50. The web browser can't do write without response GATT messages which breaks things. It might be possible to install [BlueZ 5.54 from debian backports](https://packages.debian.org/buster-backports/bluez). Not sure what this would do since raspberrypi.org ships patched versions of BlueZ and this wouldn't have those patches. It could also be possible to install the raspberrypi.org patched version of BlueZ 5.55 from Bullseye.

**Reply by scotus-49** [reply]
I completed the firmware update for my MoveHub on a Win 10 computer, and had no problems, with the download/flashing completed in about 4 mins. 

Back on my Raspberry Pi 4B, running Pybricks Code in a Web Bluetooth enabled Chromium browser, the MoveHub was recognised and paired. I have been able to download and run code OK eg. using **battery.voltage** and **battery.current** with the print command fed back the corresponding numbers to the console.

The Raspberry Pi OS updates have indeed reached 5.50 for BlueZ, but I was unable to install a version 5.54 from bullseye over this because of unmet dependencies which could not be resolved.

I will press on testing my two hubs (Move and Inventor) using Pybricks Code on the Raspi. Is there anything in particular that would be useful to check out?

**Reply by dlech** [reply]
The first post here has lots of ideas about things to test.

**Reply by scotus-49** [reply]
Yes, it's fairly comprehensive - I'm starting now designing and building some new robots, so that  I can work meaningfully through your list.

Once again - I have to say that you both have done a remarkable job with Pybricks. Thank you.

**Comment by JorgePe** [comment]
Congratulations on this major milestone.

I have flashed 3 TechnicHub without issues. Ubuntu 21.04, Chrome 90.0.4430.212 64 bit

Also been using Jupyter Notebook and pybricksdev  - 2 Control+ hubs (only issue with python already above 3.8).

Using keyboard.poll and stdin to send keys to the hubs (ok, not sticking to documentation). It stopped working once (no keys were been received) not sure why, a refresh on Chrome fixed it.

Program length is around 300 lines but it's not a large program (lot's of if/then). 4 large motors on each hub, using track_target and run_until_stalled... only issue I have, with 2 motors, is 'zero' position not correct at start. I am still trying to understand if it's a motor issue (there is an issue with 51515 smaller motors, not aware if these large motors are also affected and the offset is much larger) or just my fault.

Once batteries low the Hub shuts off when using all motors. Expected, I believe, these motors are hungry.

**Reply by laurensvalk** [reply]
Thanks for reminding us about the motor offset issue. I believe this is what that initialization routine in the Robot Inventor app is for. We are not doing this yet, and we'd have to dig into how this works.

**Reply by laurensvalk** [reply]
I just checked, and this should go away if you first run the motor update routine in the official app. Can you try this?

**Reply by JorgePe** [reply]
I was planning that... just need to find my Charlie in the middle of the mess (1 year if confinement and now major repair works, everything in boxes pilled everywhere possible). I tried Control+ App hoping that it could do the same on TechnicHub but it probably doesn't and I don't have a configuration acceptable by the App (silly thing)

**Reply by laurensvalk** [reply]
I just ran it for the very first time as well. Looks like you need one Inventor Hub with all four motors attached, but in no particular build.

**Reply by JorgePe** [reply]
I am trying to get a second Inventor Hub. But I don't need a whole 51515 and just the hub is hard to find. I'll report in a few days.

**Reply by JorgePe** [reply]
Hmm... I might have a bad lot of motors.
These are the Large Angular motors - the grey from the Volvo set and the Blue/White from SPIKE.

App is useless for calibration, requires a MINDSTORMS motor and doesn't recognize the Large Angular motors as so (a grey on Port A and a Blue/White on port C). But it can read their position - the problematic on Port A is shown at 261º and the other (good) is shown at 4º.
When using Pybricks with Technic Hub the problematic motor reads -96º at zero position, that's just 3º different of what LEGO App reads. I was going crazy with my two of my bagpiper fingers jumping to silly positions with no reason.

Almost 100º offset is crazy. Got them on  Bricklink but no reason to suspect of damage until now.

**Reply by dlech** [reply]
I started a new issue for this: https://github.com/pybricks/support/issues/343

**Comment by scotus-49** [comment]
### MoveHub on Ubuntu 21.04 with Google Chrome

_Error Message: Connected to hub but failed to get Device Information service.
Ensure that you are using the most recent firmware.
If the problem persists, try removing the "Pybricks Hub" device in your OS Bluetooth settings, then try connecting again._

I followed the advice, removed Pybricks Hub several times - same error message.
The same hub connects OK on Win 10 and a Samsung Tab both using Google Chrome.

My InventorHub works fine with the same setup (Ubuntu/Chrome).

**Reply by JorgePe** [reply]
Did you managed to install Pybricks firmware? I am also using Ubuntu 21.04 and Chrome and had no problems, tried both Move Hub "Status light examples" with success.

**Reply by scotus-49** [reply]
Firmware update stalled and did not work on Ubuntu, so I moved to a Win 10 machine and reinstalled the firmware without a hitch. I tested the hub on Win 10/Chrome and everything works as expected.

Glad to hear that the MoveHub can work on Ubuntu. I've just checked the MoveHub again using Chrome (Version 90.0.4430.212 (Official Build) (64-bit)), with experimental flags/bluetooth enabled, and the same error messages again.

**Reply by JorgePe** [reply]
Same version of Chrome here... Is it the same computer (dual booting between Win10/Ubuntu) or another one?
Can you show 'hciconfig -a' result on command line?

hciconfig -a
hci0:	Type: Primary  Bus: USB
	BD Address: 84:5C:F3:C6:8F:30  ACL MTU: 1021:4  SCO MTU: 96:6
	UP RUNNING PSCAN 
	RX bytes:1004601 acl:45391 sco:0 events:34370 errors:0
	TX bytes:978104 acl:32347 sco:0 commands:853 errors:0
	Features: 0xbf 0xfe 0x0f 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH SNIFF 
	Link mode: SLAVE ACCEPT 
	Name: 'xxxxxxxxxx'
	Class: 0x2c010c
	Service Classes: Rendering, Capturing, Audio
	Device Class: Computer, Laptop
	HCI Version: 5.2 (0xb)  Revision: 0x2279
	LMP Version: 5.2 (0xb)  Subversion: 0x2279
	Manufacturer: Intel Corp. (2)

**Reply by scotus-49** [reply]
I have a dedicated Linux box, and running ```hciconfig -a``` gives:
```
hci0:	Type: Primary  Bus: USB
	BD Address: 00:E1:8C:5A:95:05  ACL MTU: 1021:5  SCO MTU: 96:6
	UP RUNNING PSCAN ISCAN 
	RX bytes:37804 acl:596 sco:0 events:1863 errors:0
	TX bytes:53961 acl:265 sco:0 commands:1003 errors:0
	Features: 0xff 0xfe 0x0f 0xfe 0xdb 0xff 0x7b 0x87
	Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3 
	Link policy: RSWITCH HOLD SNIFF 
	Link mode: SLAVE ACCEPT 
	Name: 'ChromeLinux_AB0D'
	Class: 0x3c0104
	Service Classes: Rendering, Capturing, Object Transfer, Audio
	Device Class: Computer, Desktop workstation
	HCI Version: 4.2 (0x8)  Revision: 0x1000
	LMP Version: 4.2 (0x8)  Subversion: 0x1000
	Manufacturer: Intel Corp. (2)
```
I note that you have HCI 5.2 compared to my 4.2. 
Running ```$ bluetoothd --version``` shows 5.56. at the moment.

**Reply by JorgePe** [reply]
5.56 also. Our software is probably on pair.
HCI 4.2 should be enough. You don't have another USB BT 4.x adapter that you can try, setting 'hci0' off to force Chrome to use just the oter one?
Probably also better if you open a support issue here in pybricks

**Reply by scotus-49** [reply]
I have a spare Bluetooth dongle, and plugged it in. However, I'm unsure of the command to shut down hci0

**Reply by dlech** [reply]
Also see the Bluetooth Troubleshooting Guide: https://github.com/pybricks/support/discussions/270

**Reply by scotus-49** [reply]
Developer tools Console gives only "Web Bluetooth is experimental on this platform" with a link to a 404 page not found on Github.
The Device Log is a bit longer - what can I do with it?

EDIT: I have just managed to connect my InventorHub successfully on the same software. The developer tools console on Chrome this time read:
```
Hub firmware version: 3.0.0b4           sagas.ts:202 
Pybricks protocol version: 1.0.0        sagas.ts:230
```

**Reply by dlech** [reply]
Open a new issue describing in detail how to reproduce the problem and include the log.

**Reply by scotus-49** [reply]
Well, I understand that Web Bluetooth is still experimental on Linux, so I'm not going to push this. It works on Windows, and that will do for now. Thanks.

**Comment by JorgePe** [comment]
"Try saving and loading code files from your computer."
Saving and loading works fine.
Noobie with Chrome... I was finding strange all my programs being named "main.py".  Changed Chrome settings (Advanced, Downloads) to ask where to save each file before downloading :D

**Reply by scotus-49** [reply]
Thanks - so simple. I forgot about this, and had a folder full of main(xx).py files.

**Comment by scotus-49** [comment]
### MoveHub:  Firmware include current program | Win 10/Chrome

This worked for me with a short (25 line) program. Useful for testing hardware.

**Comment by JorgePe** [comment]
"Try including a program in the firmware via the settings menu."

Working with 2x Technic Hubs. Very useful.

**Comment by scotus-49** [comment]
### Pybricks Code text input flipping between insert and replace mode

While writing code for Pybricks using Win 10 and Chrome, I found yesterday that text input was flipping between insert and replace mode, and could not work out how that had happened. Is this a feature triggered by some accidental key combinations that I made, or something else?

**Reply by dlech** [reply]
This happens when you press the <kbd>Insert</kbd> key on your keyboard.

**Reply by scotus-49** [reply]
Thanks - nice to know. Except - I was nowhere near the insert key. I never use it. 

I have a US keyboard for my Linux box, and a German one for Windows, and the insert key for both is out of reach of my small fingers when I'm typing.

Do I understand that if I find myself in replace mode I should press the insert key to get out? That would be nice.

A small extra point regarding Pybricks code - I used it on my Samsung tablet last week, and some of the tooltips for the top bar stayed visible, making it difficult to type underneath. That's  Android 7.0, with an up-to-date Chrome browser. My tab paired nicely with the MoveHub, but the "Disconnect Bluetooth" tooltip, and sometimes "Download and run this program (F5)" stayed on screen, blotting out the top of the text.

**Reply by dlech** [reply]
> some of the tooltips for the top bar stayed visible, making it difficult to type underneath

We would probably have to switch to a different UI toolkit that is more touch-screen friendly to fix this. Also see https://github.com/pybricks/pybricks-code/issues/275.

**Reply by dlech** [reply]
> Do I understand that if I find myself in replace mode I should press the insert key to get out? That would be nice.

I would expect so. I don't know of any other way to trigger it.

**Reply by dlech** [reply]
Google says <kbd>Shift</kbd>+<kbd>Num Pad 0</kbd> will do it too.

**Reply by scotus-49** [reply]
Super. Thanks again.

**Comment by scotus-49** [comment]
### Another Try on my Raspberry Pi 4B: Pybricks Connected and ran code for built-in motors

I have a small biped robot using the two motors built-in to the MoveHub. Using the Raspbian release of Chromium, it paired, and ran a simple walking code. Details from the developer tools console:
```
Hub firmware version: 3.0.0c1
main.f60a9339.chunk.js:1 Pybricks protocol version: 1.0.0
main.f60a9339.chunk.js:1 Downloading 716 bytes
```
**System**: Raspbian GNU/Linux | Version 10 (Buster)
**Browser**: Chromium Version 88.0.4324.187 (Official Build) Built on Raspbian , running on Raspbian 10 (32-bit)
**Pybricks Code Version**: Pybricks Code v1.0.0-rc.2

**Main Bug**: When I disconnect the hub, reconnecting requires the browser to be closed and re-opened. 

**My Basic Code**:
```
from pybricks.hubs import MoveHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Port, Stop
from pybricks.tools import wait

myHub = MoveHub()

# Initialize motors
motor_left = Motor(Port.A)
motor_right = Motor(Port.B)

def walk_forward():
    motor_right.run_angle(-250, 360, wait=False)
    motor_left.run_angle(250, 360, wait=False)
    while not motor_right.control.done() or not motor_left.control.done():
        wait(10)
    wait(1000)

while True:
    walk_forward()
```

**Reply by scotus-49** [reply]
I understand that this really should not have worked, but to a certain extent, it did, and I was delighted to see my MoveHub based walker working. Daniel Benedettelli's excellent book on the **Lego Boost** has shown how much potential is available with this set.

I'm hoping that Pybricks could eventually be fully working on the Raspberry Pi. This obviously could open it up to one of the world's largest coding communities, currently stoking up interest in **micropython** for the new Raspberry Pi **Pico** board. 

My Pi 4B 2MB, with a basic system install runs **[node-poweredup](https://github.com/nathankellenicki/node-poweredup)** on my MoveHub, using the standard Lego firmware. No problems there other than writing Javascript instead of Python.

I have noted David Lechner's recent comments (pybricksdev on Raspberry Pi #342) about future developments of the Raspberry Pi OS, and will keep a lookout for any news.
